L44-python-rich-status.md

```diff
@@ -0,0 +1,474 @@
+
+The one primary use case of this feature is Google Cloud API. The Google Cloud API needs it to transport rich status from the server to the client. Also, they are the owner of those proto files.
+
+So, the gRPC team introduced a new internal trailing metadata entry `grpc-status-details-bin` to serve this purpose. It takes in serialized proto `google.rpc.status.Status` message binary and transmits as binary metadata. However, since gRPC is a ProtoBuf-agnostic framework, it can't enforce either the content of the binary data be set in the trailing metadata entry, nor the consistency of status code/status message with the additional status detail proto message. The result of this unsolvable problem is that for three major implementations of gRPC we have, each of them is slightly different about this behavior.
+
+### Behavior Different Across Implementations
+
+C++ implementation tolerates anything to be put in the `grpc-status-details-bin` entry; Java checks if the code of the proto status is matching the status code of the RPC call; Golang enforces both code and message of the proto status, and the RPC call be the same.
+
+So, which paradigm should Python follow?
+
+### The Proto of Status
+
+The proto of status is well-defined and used in many frameworks. Here is the definition of `Status`, for full version see [status.proto](https://github.com/googleapis/api-common-protos/blob/master/google/rpc/status.proto).
+```proto
+message Status {
+  int32 code = 1;
+  string message = 2;
+  repeated google.protobuf.Any details = 3;
+}
+```
+
+### Current Python API About RPC Status
+
+```Python
+# Client side
+stub = ...Stub(channel)
+try:
+    resp = stub.ARpcCall(...)
+except grpc.RpcError as rpc_error:
+    code = rpc_error.code()
+    details = rpc_error.details()
+    # Unable to get rich status
+```
+
+```Python
+# Server side
+def ...Servicer(...):
+
+    def ARPCCall(request, context):
+        context.set_code(...)
+        context.set_details(...)
+        # Unable to set rich status
+
+```
+
+## Proposal
+
+### A New Public Interface: Status
+
+This class is used to describe the status of an RPC.
+
+The name of the class is `Status`, because it is shorter and cleaner than `RpcStatus` or `ServerRpcStatus`. In the future, we might want to utilize it at client side as well.
+
+The metadata field of `Status` is `trailing_metadata` instead of `metadata`. The `Status` should be the final status of an RPC, so it should assist developers to get information that only accessible at the end of the RPC. Also, if we want to support it at client side, we need to separate `initial_metadata` and `trailing_metadata`.
```

Comment:
I don't understand the last sentence about supporting this at client side, I think it can just be deleted. On the client side the status would not included initial_metadata...but it doesn't matter for this gRFC in either case.

Reply:
The sentence is removed. I meant there have to be two fields to support both server-side and client-side.